include/Cell.h

```diff
@@ -22,64 +22,68 @@
 
 class Cell {
 public:
+  typedef std::vector<std::weak_ptr<Cell>> WeakCells;
+  typedef std::shared_ptr<Cell>            SCell;
+  typedef std::weak_ptr<Cell>              WCell;
+
+public:
   // Constructors, main initialisation is with two kd hits
   Cell() { m_weight = 0; }
   Cell(const Cell&) = delete;
   Cell& operator=(const Cell&) = delete;
 
-  Cell(KDCluster* hit1, KDCluster* hit2)
+  Cell(SKDCluster const& hit1, SKDCluster const& hit2)
       : m_weight(0),
         m_gradient((hit2->getV() - hit1->getV()) / (hit2->getU() - hit1->getU())),
         m_gradientRZ((hit2->getRadius() - hit1->getRadius()) / (hit2->getZ() - hit1->getZ())),
         m_start(hit1),
         m_end(hit2) {}
 
   // Destructor
-  virtual ~Cell() {
-    m_from.clear();
-    m_to.clear();
-  }
+  virtual ~Cell() {}
 
   // Weight of the cell (first cell in a chain has weight 0, and each subsequent link has weight +1)
-  int  getWeight() { return m_weight; }
+  int  getWeight() const { return m_weight; }
   void setWeight(int weight) { m_weight = weight; }
 
   // Gradient of the cell connecting two hits
-  double getGradient() { return m_gradient; }
+  double getGradient() const { return m_gradient; }
   void setGradient(double gradient) { m_gradient = gradient; }
-  double                  getGradientRZ() { return m_gradientRZ; }
+  double                  getGradientRZ() const { return m_gradientRZ; }
 
   // Angle between two cells. This is assumed to be less than 90 degrees
-  double getAngle(Cell* cell2) {
+  double getAngle(SCell cell2) const { return getAngle(cell2.get()); }
+  double getAngle(Cell* cell2) const {
     return fabs(std::atan((cell2->getGradient() - m_gradient) / (1 + m_gradient * cell2->getGradient())));
   }
-  double getAngleRZ(Cell* cell2) {
+  double getAngleRZ(SCell cell2) const { return getAngleRZ(cell2.get()); }
+  double getAngleRZ(Cell* cell2) const {
     return fabs(std::atan((cell2->getGradientRZ() - m_gradientRZ) / (1 + m_gradientRZ * cell2->getGradientRZ())));
   }
 
   // Start and end points of the cell
-  KDCluster* getStart() { return m_start; }
-  KDCluster* getEnd() { return m_end; }
+  SKDCluster getStart() const { return m_start; }
+  SKDCluster getEnd() const { return m_end; }
 
   // Increment the cell weight (usually if the chain length is extended upstream of this cell)
-  void update(Cell* cell2) {
+  void update(SCell const& cell2) {
     if ((cell2->getWeight() + 1) > m_weight)
       m_weight = cell2->getWeight() + 1;
   }
 
   // The cell has a memory of all cells that connect to it, and all cells that it connects to. If several cells point to this
   // cell, then the weight taken from the highest weighted of those (longest chain)
-  void setFrom(Cell* cell2) {
-    m_from.push_back(cell2);
+  void setFrom(SCell const& cell2) {
```

Comment:
what is the purpose of the const here?

Reply:
This signals that cell2 is not modified in the function call, as we are dealing with a shared_ptr that we copy we would probably have the same performance if we just pass by copy here.